cats/cats-redis/src/main/java/com/netflix/spinnaker/cats/redis/cluster/CachingPodsObserver.java

```diff
@@ -0,0 +1,131 @@
+import com.netflix.spinnaker.kork.jedis.RedisClientDelegate;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CachingPodsObserver implements ShardingFilter, Runnable {
+
+  private static final Logger logger = LoggerFactory.getLogger(CachingPodsObserver.class);
+  private int podCount = 0;
+  private int podIndex = -1;
+  private static final String REPLICA_SSET_KEY = "clouddriver:caching:replicas";
+  private static final String CORE_PROVIDER =
+      "com.netflix.spinnaker.clouddriver.core.provider.CoreProvider";
+  private final RedisClientDelegate redisClientDelegate;
+  private final NodeIdentity nodeIdentity;
+  private final long replicaKeyTtl;
+  // this script adds or updates a unique id as a member of a sorted set with score equal to current
+  // time plus
+  // sharding.replica-key-ttl-seconds, deletes the members having scores less than current time(ms)
+  // and finally
+  // fetches list of all members of the sorted set which represent the live caching pods
+  private static final String HEARTBEAT_REFRESH_SCRIPT =
+      "redis.call('zadd', KEYS[1], ARGV[1], ARGV[2])"
+          + " redis.call('zremrangebyscore', KEYS[1], '-inf', ARGV[3])"
+          + " return redis.call('zrange', KEYS[1], '0', '-1')";
+
+  public CachingPodsObserver(
+      RedisClientDelegate redisClientDelegate,
+      NodeIdentity nodeIdentity,
+      DynamicConfigService dynamicConfigService) {
+    this.redisClientDelegate = redisClientDelegate;
+    this.nodeIdentity = nodeIdentity;
+    long observerIntervalSeconds =
+        dynamicConfigService.getConfig(Integer.class, "sharding.observer-interval-seconds", 30);
+    replicaKeyTtl =
+        dynamicConfigService.getConfig(Integer.class, "sharding.replica-key-ttl-seconds", 60);
+    ScheduledExecutorService podsObserverExecutorService =
+        Executors.newSingleThreadScheduledExecutor(
+            new ThreadFactoryBuilder()
+                .setNameFormat(CachingPodsObserver.class.getSimpleName() + "-%d")
+                .build());
+    podsObserverExecutorService.scheduleAtFixedRate(
+        this, 0, observerIntervalSeconds, TimeUnit.SECONDS);
+    refreshHeartbeat();
+    logger.info("Account based sharding is enabled for all caching pods.");
+  }
+
+  @Override
+  public void run() {
+    try {
+      refreshHeartbeat();
+    } catch (Exception e) {
+      logger.error("Failed to manage replicas heartbeat", e);
+    }
+  }
+
+  private void refreshHeartbeat() {
+    String now = String.valueOf(System.currentTimeMillis());
+    String expiry = String.valueOf(System.currentTimeMillis() + replicaKeyTtl);
+    Object evalResponse =
+        redisClientDelegate.withScriptingClient(
```

Comment:
is it worthwhile to add retries around it, similar to how it is [elsewhere in the code](https://github.com/spinnaker/clouddriver/blob/master/clouddriver-core/src/main/java/com/netflix/spinnaker/clouddriver/data/task/jedis/RedisTaskRepository.java#L118-L124)?

Reply:
This is not a critical operation and occasional failures do not affect the functionality. I feel adding retires is an overkill but if it is proven otherwise when adopted, I can certainly add the retry logic.